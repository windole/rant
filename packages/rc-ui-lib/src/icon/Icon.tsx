import React, { useContext, useMemo } from 'react';
import classnames from 'classnames';
import { addUnit } from '../utils';
import { IconProps } from './PropsType';
import Badge from '../badge';
import ConfigProviderContext from '../config-provider/ConfigProviderContext';

function isImage(name?: string): boolean {
  return name ? name.indexOf('/') !== -1 : false;
}

const Icon: React.FC<IconProps> = (props) => {
  const { iconPrefix } = useContext(ConfigProviderContext);

  const { dot, tag, name, className, onClick, onTouchStart } = props;

  const imageIcon = isImage(name);

  const classPrefix = useMemo(
    () => props.classPrefix || iconPrefix,
    [props.classPrefix, iconPrefix],
  );

  return (
    <Badge
      dot={dot}
      tag={tag}
      className={classnames(className, classPrefix, imageIcon ? '' : `${classPrefix}-${name}`)}
      style={{
        color: props.color,
        fontSize: addUnit(props.size),
        ...props.style,
      }}
      onClick={onClick}
      onTouchStart={onTouchStart}
      {...props.badge}
    >
      {props?.children}
      {imageIcon && <img className={classnames('rc-icon__image')} src={name} alt={name} />}
    </Badge>
  );
};

Icon.defaultProps = {
  tag: 'i',
};

export default Icon;
