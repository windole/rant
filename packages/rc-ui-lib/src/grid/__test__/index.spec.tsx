import * as React from 'react';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import Icon from '../../icon';
import Grid from '..';

describe('Grid', () => {
  let wrapper;
  afterEach(() => {
    wrapper.unmount();
  });

  it('should render square grid with gutter correctly', async () => {
    wrapper = mount(
      <Grid square columnNum={2} gutter={10}>
        <Grid.Item />
        <Grid.Item />
        <Grid.Item />
      </Grid>,
    );
    expect(toJson(wrapper.html())).toMatchSnapshot();
  });

  it('should render correctly with index prop', async () => {
    wrapper = mount(
      <Grid gutter={10} columnNum={4}>
        {Array.from({ length: 8 }, (_, i) => (
          <Grid.Item key={i} icon="photo-o" text="文字" />
        ))}
      </Grid>,
    );
    expect(toJson(wrapper.html())).toMatchSnapshot();
  });

  it('should change icon size when using icon-size prop', async () => {
    wrapper = mount(
      <Grid iconSize="10">
        <Grid.Item icon="success" />
      </Grid>,
    );
    expect(wrapper.find('i').props().style.fontSize).toEqual('10px');
  });

  it('should change icon color when using icon-color prop', async () => {
    wrapper = mount(
      <Grid iconSize="10">
        <Grid.Item icon="success" iconColor="red" />
      </Grid>,
    );
    expect(wrapper.find('i').props().style.color).toEqual('red');
  });

  it('should render icon correctly', async () => {
    wrapper = mount(
      <Grid>
        <Grid.Item badge={{ content: 1 }} icon="success" iconColor="red" />
      </Grid>,
    );
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should render correctly when icon is ReactNode', async () => {
    wrapper = mount(
      <Grid>
        <Grid.Item icon={<Icon name="shop-o" />} />
      </Grid>,
    );
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should render ".rc-grid-item__content--reverse" class when using reverse prop', async () => {
    wrapper = mount(
      <Grid reverse>
        <Grid.Item />
      </Grid>,
    );
    expect(
      wrapper.find('.rc-grid-item__content').hasClass('rc-grid-item__content--reverse'),
    ).toBeTruthy();
  });

  it('should render correctly when using text prop', async () => {
    wrapper = mount(
      <Grid>
        <Grid.Item icon="photo-o" text="hello,world!" />
      </Grid>,
    );
    expect(toJson(wrapper.html())).toMatchSnapshot();
  });

  it('should render correctly when text prop is ReactNode', async () => {
    wrapper = mount(
      <Grid>
        <Grid.Item icon="photo-o" text={React.createElement('div', {}, 'world')} />
      </Grid>,
    );
    expect(toJson(wrapper.html())).toMatchSnapshot();
  });

  it('should render correctly when using iconPrefix prop', async () => {
    wrapper = mount(
      <Grid>
        <Grid.Item icon="photo-o" iconPrefix="my-icon" text="hello,world!" />
      </Grid>,
    );
    expect(wrapper.find('i').hasClass('my-icon-photo-o')).toBeTruthy();
  });

  it('should render correctly when using iconColor prop', async () => {
    wrapper = mount(
      <Grid>
        <Grid.Item icon="photo-o" iconColor="red" />
      </Grid>,
    );
    expect(wrapper.find('i').props().style.color).toEqual('red');
  });

  it('should render correctly when using contentClassName prop', async () => {
    wrapper = mount(
      <Grid>
        <Grid.Item icon="photo-o" text="picture" contentClassName="my-content" />
      </Grid>,
    );
    expect(wrapper.find('.my-content').exists()).toBeTruthy();
  });

  it('should render correctly when using contentStyle prop', async () => {
    wrapper = mount(
      <Grid>
        <Grid.Item icon="photo-o" text="picture" contentStyle={{ color: 'red' }} />
      </Grid>,
    );
    expect(wrapper.find('.rc-grid-item__content').props().style.color).toEqual('red');
  });

  it('should emit click event correctly', async () => {
    const onClick = jest.fn();
    wrapper = mount(
      <Grid>
        <Grid.Item icon="photo-o" text="picture" onClick={onClick} />
      </Grid>,
    );
    wrapper.find('.rc-grid-item__content').simulate('click');

    expect(onClick).toHaveBeenCalled();
  });

  it('should render correctly with children', async () => {
    wrapper = mount(
      <Grid>
        <Grid.Item icon="photo-o">hello,world!</Grid.Item>
      </Grid>,
    );

    expect(toJson(wrapper.html())).toMatchSnapshot();
  });
});
