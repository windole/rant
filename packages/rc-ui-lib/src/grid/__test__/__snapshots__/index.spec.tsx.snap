// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Grid should render correctly when icon is ReactNode 1`] = `"<div class=\\"rc-grid rc-hairline--top\\"><div class=\\"rc-grid-item\\" style=\\"flex-basis: 25%;\\"><div class=\\"rc-grid-item__content rc-grid-item__content--center rc-hairline\\"><div class=\\"rc-badge__wrapper\\"><i class=\\"rc-badge__wrapper van-icon van-icon-shop-o\\"></i></div></div></div></div>"`;

exports[`Grid should render correctly when text prop is ReactNode 1`] = `
<Grid
  border={true}
  center={true}
  columnNum={4}
>
  <div
    className="rc-grid rc-hairline--top"
    style={
      Object {
        "paddingLeft": undefined,
      }
    }
  >
    <GridItem
      icon="photo-o"
      index={0}
      key=".0"
      parent={
        Object {
          "border": true,
          "center": true,
          "columnNum": 4,
        }
      }
      text={
        <div>
          world
        </div>
      }
    >
      <div
        className="rc-grid-item"
        style={
          Object {
            "flexBasis": "25%",
          }
        }
      >
        <div
          className="rc-grid-item__content rc-grid-item__content--center rc-hairline"
        >
          <Icon
            className="rc-grid-item__icon"
            name="photo-o"
            tag="i"
          >
            <Badge
              className="rc-grid-item__icon van-icon van-icon-photo-o"
              showZero={true}
              style={
                Object {
                  "color": undefined,
                  "fontSize": undefined,
                }
              }
              tag="i"
            >
              <i
                className="rc-badge__wrapper rc-grid-item__icon van-icon van-icon-photo-o"
                style={
                  Object {
                    "color": undefined,
                    "fontSize": undefined,
                  }
                }
              />
            </Badge>
          </Icon>
          <div>
            world
          </div>
        </div>
      </div>
    </GridItem>
  </div>
</Grid>
`;

exports[`Grid should render correctly when using text prop 1`] = `
<Grid
  border={true}
  center={true}
  columnNum={4}
>
  <div
    className="rc-grid rc-hairline--top"
    style={
      Object {
        "paddingLeft": undefined,
      }
    }
  >
    <GridItem
      icon="photo-o"
      index={0}
      key=".0"
      parent={
        Object {
          "border": true,
          "center": true,
          "columnNum": 4,
        }
      }
      text="hello,world!"
    >
      <div
        className="rc-grid-item"
        style={
          Object {
            "flexBasis": "25%",
          }
        }
      >
        <div
          className="rc-grid-item__content rc-grid-item__content--center rc-hairline"
        >
          <Icon
            className="rc-grid-item__icon"
            name="photo-o"
            tag="i"
          >
            <Badge
              className="rc-grid-item__icon van-icon van-icon-photo-o"
              showZero={true}
              style={
                Object {
                  "color": undefined,
                  "fontSize": undefined,
                }
              }
              tag="i"
            >
              <i
                className="rc-badge__wrapper rc-grid-item__icon van-icon van-icon-photo-o"
                style={
                  Object {
                    "color": undefined,
                    "fontSize": undefined,
                  }
                }
              />
            </Badge>
          </Icon>
          <span
            className="rc-grid-item__text"
          >
            hello,world!
          </span>
        </div>
      </div>
    </GridItem>
  </div>
</Grid>
`;

exports[`Grid should render correctly with children 1`] = `
<Grid
  border={true}
  center={true}
  columnNum={4}
>
  <div
    className="rc-grid rc-hairline--top"
    style={
      Object {
        "paddingLeft": undefined,
      }
    }
  >
    <GridItem
      icon="photo-o"
      index={0}
      key=".0"
      parent={
        Object {
          "border": true,
          "center": true,
          "columnNum": 4,
        }
      }
    >
      <div
        className="rc-grid-item"
        style={
          Object {
            "flexBasis": "25%",
          }
        }
      >
        <div
          className="rc-grid-item__content rc-grid-item__content--center rc-hairline"
        >
          hello,world!
        </div>
      </div>
    </GridItem>
  </div>
</Grid>
`;

exports[`Grid should render correctly with index prop 1`] = `
<Grid
  border={true}
  center={true}
  columnNum={4}
  gutter={10}
>
  <div
    className="rc-grid"
    style={
      Object {
        "paddingLeft": "10px",
      }
    }
  >
    <GridItem
      icon="photo-o"
      index={0}
      key=".$0"
      parent={
        Object {
          "border": true,
          "center": true,
          "columnNum": 4,
          "gutter": 10,
        }
      }
      text="文字"
    >
      <div
        className="rc-grid-item"
        style={
          Object {
            "flexBasis": "25%",
            "paddingRight": "10px",
          }
        }
      >
        <div
          className="rc-grid-item__content rc-grid-item__content--center rc-grid-item__content--surround rc-hairline"
        >
          <Icon
            className="rc-grid-item__icon"
            name="photo-o"
            tag="i"
          >
            <Badge
              className="rc-grid-item__icon van-icon van-icon-photo-o"
              showZero={true}
              style={
                Object {
                  "color": undefined,
                  "fontSize": undefined,
                }
              }
              tag="i"
            >
              <i
                className="rc-badge__wrapper rc-grid-item__icon van-icon van-icon-photo-o"
                style={
                  Object {
                    "color": undefined,
                    "fontSize": undefined,
                  }
                }
              />
            </Badge>
          </Icon>
          <span
            className="rc-grid-item__text"
          >
            文字
          </span>
        </div>
      </div>
    </GridItem>
    <GridItem
      icon="photo-o"
      index={1}
      key=".$1"
      parent={
        Object {
          "border": true,
          "center": true,
          "columnNum": 4,
          "gutter": 10,
        }
      }
      text="文字"
    >
      <div
        className="rc-grid-item"
        style={
          Object {
            "flexBasis": "25%",
            "paddingRight": "10px",
          }
        }
      >
        <div
          className="rc-grid-item__content rc-grid-item__content--center rc-grid-item__content--surround rc-hairline"
        >
          <Icon
            className="rc-grid-item__icon"
            name="photo-o"
            tag="i"
          >
            <Badge
              className="rc-grid-item__icon van-icon van-icon-photo-o"
              showZero={true}
              style={
                Object {
                  "color": undefined,
                  "fontSize": undefined,
                }
              }
              tag="i"
            >
              <i
                className="rc-badge__wrapper rc-grid-item__icon van-icon van-icon-photo-o"
                style={
                  Object {
                    "color": undefined,
                    "fontSize": undefined,
                  }
                }
              />
            </Badge>
          </Icon>
          <span
            className="rc-grid-item__text"
          >
            文字
          </span>
        </div>
      </div>
    </GridItem>
    <GridItem
      icon="photo-o"
      index={2}
      key=".$2"
      parent={
        Object {
          "border": true,
          "center": true,
          "columnNum": 4,
          "gutter": 10,
        }
      }
      text="文字"
    >
      <div
        className="rc-grid-item"
        style={
          Object {
            "flexBasis": "25%",
            "paddingRight": "10px",
          }
        }
      >
        <div
          className="rc-grid-item__content rc-grid-item__content--center rc-grid-item__content--surround rc-hairline"
        >
          <Icon
            className="rc-grid-item__icon"
            name="photo-o"
            tag="i"
          >
            <Badge
              className="rc-grid-item__icon van-icon van-icon-photo-o"
              showZero={true}
              style={
                Object {
                  "color": undefined,
                  "fontSize": undefined,
                }
              }
              tag="i"
            >
              <i
                className="rc-badge__wrapper rc-grid-item__icon van-icon van-icon-photo-o"
                style={
                  Object {
                    "color": undefined,
                    "fontSize": undefined,
                  }
                }
              />
            </Badge>
          </Icon>
          <span
            className="rc-grid-item__text"
          >
            文字
          </span>
        </div>
      </div>
    </GridItem>
    <GridItem
      icon="photo-o"
      index={3}
      key=".$3"
      parent={
        Object {
          "border": true,
          "center": true,
          "columnNum": 4,
          "gutter": 10,
        }
      }
      text="文字"
    >
      <div
        className="rc-grid-item"
        style={
          Object {
            "flexBasis": "25%",
            "paddingRight": "10px",
          }
        }
      >
        <div
          className="rc-grid-item__content rc-grid-item__content--center rc-grid-item__content--surround rc-hairline"
        >
          <Icon
            className="rc-grid-item__icon"
            name="photo-o"
            tag="i"
          >
            <Badge
              className="rc-grid-item__icon van-icon van-icon-photo-o"
              showZero={true}
              style={
                Object {
                  "color": undefined,
                  "fontSize": undefined,
                }
              }
              tag="i"
            >
              <i
                className="rc-badge__wrapper rc-grid-item__icon van-icon van-icon-photo-o"
                style={
                  Object {
                    "color": undefined,
                    "fontSize": undefined,
                  }
                }
              />
            </Badge>
          </Icon>
          <span
            className="rc-grid-item__text"
          >
            文字
          </span>
        </div>
      </div>
    </GridItem>
    <GridItem
      icon="photo-o"
      index={4}
      key=".$4"
      parent={
        Object {
          "border": true,
          "center": true,
          "columnNum": 4,
          "gutter": 10,
        }
      }
      text="文字"
    >
      <div
        className="rc-grid-item"
        style={
          Object {
            "flexBasis": "25%",
            "marginTop": "10px",
            "paddingRight": "10px",
          }
        }
      >
        <div
          className="rc-grid-item__content rc-grid-item__content--center rc-grid-item__content--surround rc-hairline"
        >
          <Icon
            className="rc-grid-item__icon"
            name="photo-o"
            tag="i"
          >
            <Badge
              className="rc-grid-item__icon van-icon van-icon-photo-o"
              showZero={true}
              style={
                Object {
                  "color": undefined,
                  "fontSize": undefined,
                }
              }
              tag="i"
            >
              <i
                className="rc-badge__wrapper rc-grid-item__icon van-icon van-icon-photo-o"
                style={
                  Object {
                    "color": undefined,
                    "fontSize": undefined,
                  }
                }
              />
            </Badge>
          </Icon>
          <span
            className="rc-grid-item__text"
          >
            文字
          </span>
        </div>
      </div>
    </GridItem>
    <GridItem
      icon="photo-o"
      index={5}
      key=".$5"
      parent={
        Object {
          "border": true,
          "center": true,
          "columnNum": 4,
          "gutter": 10,
        }
      }
      text="文字"
    >
      <div
        className="rc-grid-item"
        style={
          Object {
            "flexBasis": "25%",
            "marginTop": "10px",
            "paddingRight": "10px",
          }
        }
      >
        <div
          className="rc-grid-item__content rc-grid-item__content--center rc-grid-item__content--surround rc-hairline"
        >
          <Icon
            className="rc-grid-item__icon"
            name="photo-o"
            tag="i"
          >
            <Badge
              className="rc-grid-item__icon van-icon van-icon-photo-o"
              showZero={true}
              style={
                Object {
                  "color": undefined,
                  "fontSize": undefined,
                }
              }
              tag="i"
            >
              <i
                className="rc-badge__wrapper rc-grid-item__icon van-icon van-icon-photo-o"
                style={
                  Object {
                    "color": undefined,
                    "fontSize": undefined,
                  }
                }
              />
            </Badge>
          </Icon>
          <span
            className="rc-grid-item__text"
          >
            文字
          </span>
        </div>
      </div>
    </GridItem>
    <GridItem
      icon="photo-o"
      index={6}
      key=".$6"
      parent={
        Object {
          "border": true,
          "center": true,
          "columnNum": 4,
          "gutter": 10,
        }
      }
      text="文字"
    >
      <div
        className="rc-grid-item"
        style={
          Object {
            "flexBasis": "25%",
            "marginTop": "10px",
            "paddingRight": "10px",
          }
        }
      >
        <div
          className="rc-grid-item__content rc-grid-item__content--center rc-grid-item__content--surround rc-hairline"
        >
          <Icon
            className="rc-grid-item__icon"
            name="photo-o"
            tag="i"
          >
            <Badge
              className="rc-grid-item__icon van-icon van-icon-photo-o"
              showZero={true}
              style={
                Object {
                  "color": undefined,
                  "fontSize": undefined,
                }
              }
              tag="i"
            >
              <i
                className="rc-badge__wrapper rc-grid-item__icon van-icon van-icon-photo-o"
                style={
                  Object {
                    "color": undefined,
                    "fontSize": undefined,
                  }
                }
              />
            </Badge>
          </Icon>
          <span
            className="rc-grid-item__text"
          >
            文字
          </span>
        </div>
      </div>
    </GridItem>
    <GridItem
      icon="photo-o"
      index={7}
      key=".$7"
      parent={
        Object {
          "border": true,
          "center": true,
          "columnNum": 4,
          "gutter": 10,
        }
      }
      text="文字"
    >
      <div
        className="rc-grid-item"
        style={
          Object {
            "flexBasis": "25%",
            "marginTop": "10px",
            "paddingRight": "10px",
          }
        }
      >
        <div
          className="rc-grid-item__content rc-grid-item__content--center rc-grid-item__content--surround rc-hairline"
        >
          <Icon
            className="rc-grid-item__icon"
            name="photo-o"
            tag="i"
          >
            <Badge
              className="rc-grid-item__icon van-icon van-icon-photo-o"
              showZero={true}
              style={
                Object {
                  "color": undefined,
                  "fontSize": undefined,
                }
              }
              tag="i"
            >
              <i
                className="rc-badge__wrapper rc-grid-item__icon van-icon van-icon-photo-o"
                style={
                  Object {
                    "color": undefined,
                    "fontSize": undefined,
                  }
                }
              />
            </Badge>
          </Icon>
          <span
            className="rc-grid-item__text"
          >
            文字
          </span>
        </div>
      </div>
    </GridItem>
  </div>
</Grid>
`;

exports[`Grid should render icon correctly 1`] = `"<div class=\\"rc-grid rc-hairline--top\\"><div class=\\"rc-grid-item\\" style=\\"flex-basis: 25%;\\"><div class=\\"rc-grid-item__content rc-grid-item__content--center rc-hairline\\"><i class=\\"rc-badge__wrapper rc-grid-item__icon van-icon van-icon-success\\" style=\\"color: red;\\"><div class=\\"rc-badge rc-badge--fixed\\">1</div></i></div></div></div>"`;

exports[`Grid should render square grid with gutter correctly 1`] = `
<Grid
  border={true}
  center={true}
  columnNum={2}
  gutter={10}
  square={true}
>
  <div
    className="rc-grid"
    style={
      Object {
        "paddingLeft": "10px",
      }
    }
  >
    <GridItem
      index={0}
      key=".0"
      parent={
        Object {
          "border": true,
          "center": true,
          "columnNum": 2,
          "gutter": 10,
          "square": true,
        }
      }
    >
      <div
        className="rc-grid-item rc-grid-item--square"
        style={
          Object {
            "flexBasis": "50%",
            "paddingTop": "50%",
          }
        }
      >
        <div
          className="rc-grid-item__content rc-grid-item__content--center rc-grid-item__content--square rc-grid-item__content--surround rc-hairline"
          style={
            Object {
              "bottom": "10px",
              "height": "auto",
              "right": "10px",
            }
          }
        />
      </div>
    </GridItem>
    <GridItem
      index={1}
      key=".1"
      parent={
        Object {
          "border": true,
          "center": true,
          "columnNum": 2,
          "gutter": 10,
          "square": true,
        }
      }
    >
      <div
        className="rc-grid-item rc-grid-item--square"
        style={
          Object {
            "flexBasis": "50%",
            "paddingTop": "50%",
          }
        }
      >
        <div
          className="rc-grid-item__content rc-grid-item__content--center rc-grid-item__content--square rc-grid-item__content--surround rc-hairline"
          style={
            Object {
              "bottom": "10px",
              "height": "auto",
              "right": "10px",
            }
          }
        />
      </div>
    </GridItem>
    <GridItem
      index={2}
      key=".2"
      parent={
        Object {
          "border": true,
          "center": true,
          "columnNum": 2,
          "gutter": 10,
          "square": true,
        }
      }
    >
      <div
        className="rc-grid-item rc-grid-item--square"
        style={
          Object {
            "flexBasis": "50%",
            "paddingTop": "50%",
          }
        }
      >
        <div
          className="rc-grid-item__content rc-grid-item__content--center rc-grid-item__content--square rc-grid-item__content--surround rc-hairline"
          style={
            Object {
              "bottom": "10px",
              "height": "auto",
              "right": "10px",
            }
          }
        />
      </div>
    </GridItem>
  </div>
</Grid>
`;
